### DISCUSSION TITLE
Can't configure device ID in OpenOCD unexpected device found FlashPro5

### DISCUSSION THREAD

**Discussion by Micko8** [polarfire-soc/polarfire-soc-documentation#236] [Applications and demos] [answered]
Hello all,
I want to run a bare-metal example but I need to target my device (it is not ICICLE KIT).
Project build goes well. 
I configure the debug launcher with my device MPFS250 in the config options.
But openocd is failing. You can find the entire console output attached. 
[SoftConsoleLog.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9940628/SoftConsoleLog.txt)

Basically here is the main problem : 
Info : JTAG tap: mpfs.cpu tap/device found: 0x**1**f81a1cf (mfg: 0x0e7 (GateField), part: 0xf81a, **ver: 0x1**)
Warn : JTAG tap: mpfs.cpu       UNEXPECTED: 0x1f81a1cf (mfg: 0x0e7 (GateField), part: 0xf81a, ver: 0x1)
Error: JTAG tap: mpfs.cpu expected 1 of 15: 0x0f8581cf (mfg: 0x0e7 (GateField), part: 0xf858, ver: 0x0)

As verified with Flashpro Express Device Info run, my device does export this id code 0x1f81a1cf. So so far it's not a surprise.

When I open  target/microsemi-riscv.cfg  I find my device with an idcode to 0x0f81a1cf (almost the same as  0x1f81a1cf)
MPFS250         "0x0f81a1cf 0x0f83a1cf 0x0f83e1cf"  (Even if my device is exactly MPFS250T but not present in cgf)

I have tried to custom this config file with the detected id code like it's explained here 
https://elinux.org/OpenOCD_Troubleshooting:_JTAG_Tap_Unexpected but nothing better. Openocd still waits for something my programmer is not providing with.

Could you tell me why I have this offset on the idcode (0x1f81a1cf vs 0x0f81a1cf)? What can be responsible of this ? FP5 is well installed.
I tried to learn more about JTAG TAP protocol but I can't find answers.
I think that is something to be done on my hardware because I have SW DIP-8 switch with one of position is linked to my TDO pin. I can share the schematics if one of you answers me.

Thank you

ps : my FPExpress log if needed
[FPExpresslog.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9940816/FPExpresslog.txt)

**Comment by hughbreslin** [comment, marked as answer]
Hi @Micko8 

out of pure chance I encountered a similar issue today. In my case I think the MSS in my design isn't coming out of reset (this is on a known working board and appears to be design related). You could try using SmartDebug to access MSS registers and see if you get an SCB error (this would indicate the MSS is in reset). 

If thats the case your best bet would be programming with a near empty design and verify the MSS comes out of reset and if this still fails you could open a tech support case to get some more support on this issue. If the empty design works you could add in your components again and see what causes the failure.

**Reply by Micko8** [reply]
Hello @hughbreslin ,
Thank you for sharing this. I will have a look asap and tell you my outcomes in details.

Also could you please give me your opinion about an other issue I have ?

There it is:
I can't debug with OpenOCD but I managed to run the "PolarFire SoC program non-secure boot mode 1" for my target.
I had to customize the lim linker script custom-mpfs-lim.ld and change the DIE in the external tool configuration argument (from MPFS250T-ES to MPFS250) for a successful target FlashPro5 programming.

But still there is nothing coming out from RS232 ports. I'm expecting to see E51 U54_1 displaying through UART... (mssuart-interrupt exemple project).

Do you think @hughbreslin this issue could be related to the openocd debug one I have ? I might missed something in the process.

I have also tried to program the target with my supplier Libero project after following the usual process:
Place and Route validated
Generated FPGA Array Data
Configure Design Initialization Data and Memories adding my .hex file to eNVM 
Run program action successfully.

Still nothing on uart. I have even try another RS232 --> USB converter. Would you have already experiences this issue too ? Thank you!

I keep you in touch for try with SmartDebug.

**Reply by hughbreslin** [reply]
One or two other suggestions on this - if the MSS is locking up (i.e not a trap but a literal crash where it hangs, this can occur if you try to access a memory location that doesn't exist or a peripheral with its clock disabled) it can cause a similar issue. Along with trying to debug a program that targets scratchpad when the scratchpad hasn't been configured correctly. I'm not sure if either of these help but just to give a few pointers at least. 

No output from UART could indicate its not reaching the point of being able to run the prints, if the HAL is failing somewhere in boot you might not see any output as the program isn't getting that far. What you could do to try and validate the debug works, program with boot mode 0 (this should have the MSS in a while loop), then use an attach debug configuration to try and connect to the MSS without downloading or running any code, this would eliminate the jtag set up as the issue.

**Reply by Micko8** [reply]
@hughbreslin 
Thank you I keep these advice in mind and I give my feedback to you asap. 
Cheers

**Reply by Micko8** [reply]
Hello @hughbreslin 

I have done a bit on investigation following your suggestions:
SmartDebug from my Libero project: 
   * No SCB errors. Bitstream programmed seems match with the one used by Smartdebug. No warnings
   *If I display and read SOFT_RESET_CR values, I see MMUART0 and 1 to 0x0. Out of my understanding they are not held in reset.
I rely on this tutorial : https://www.youtube.com/watch?v=7-hmSy-TdPI

But if I look at PFSOC_MSS_TOP_SYSREG_DEVICE_STATUS, CORE_UP, I can see 0x1. Maybe I should worry about this. I can't find Polarfire Register Map to by the way.

Regarding your remark on bad memory access. I'm pretty sure that I fail something with MSS Reset Vector when I try to migrate mssuart interrupt baremetal project for my target (not icicle kit). I had to replace l2lim origin address (0x08000000) with the envm one( 0x20220100) to make the boot mode 1 running successfully.

For the debug test, how can I debug without running any application code as debug configuration is targeting .elf which is out of the project build ?

Thank you for your help. I'm doing my first steps in this. Don't hesitate to tell me to refer to any documentation that I could have missed for  a better understanding  of what I'm doing.

**Reply by hughbreslin** [reply]
Hmm ok, which version of SoftConsole are you working with? If its not the latest (2022.2) could you try with that?

In terms of the linker change - I suspect this may be an issue, could you try building with the "eNVM-Scratchpad-Release" build configuration? Theres more info on the build configurations [here](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples#build-configurations). You may need to update your XML to set up a scratchpad area. 

> For the debug test, how can I debug without running any application code as debug configuration is targeting .elf which is out of the project build ?

When you debug targeting the LIM the application code is downloaded through the debugger and you don't need anything running on the MSS when doing this so its a useful test to ensure your JTAG isn't an issue.

**Reply by Micko8** [reply]
I'm already using the 2022.2 version.

I will keep going with baremetal project migration tests. Thank you.

However my biggest concern is to see nothing coming on UART when I program my board from my supplier Libero project.

As I told you I generate the bitstream, add my .hex to the eNVM memory and Run program action.
It should be straight forward even if I had to change the DIE from 250T-ES to 250T in Libero project settings (otherwise Flashpro couldn't program), unless this could insert an issue... But I don't think so.

I will rely on https://www.microsemi.com/document-portal/doc_view/136529-ug0743-polarfire-fpga-debugging-user-guide  to find out what went wrong.
But I definitely can't find the PolarfireSoc registers map to find their roles and descriptions. This link has nothing https://www.microsemi.com/document-portal/cat_view/56661-internal-documents/56758-soc/56780-other  , where should I find it ?

If you know where I should look in registers to debug, I would appreciate too ! As soon as I solve this I will try to answer to the main topic of this discussion I have opened.

**Reply by hughbreslin** [reply]
The register map is available in the [documentation section](https://www.microchip.com/en-us/products/fpgas-and-plds/system-on-chip-fpgas/polarfire-soc-fpgas#Documentation) of the PolarFire SoC product page, have a look under the "Data Sheets" section. 

There are no registers jumping to mind that could be used for this, you could check the status registers in the "pfsoc_mss_top_sysreg" and see if theres anything jumping out there. 

On the UART issue I suspect there may be an issue with the linker used for the hex but I may be wrong. Good luck with your testing.
